File: packages/js/core/src/types/UriResolver.ts
import { Uri } from ".";
import { IUriResolutionContext, UriPackageOrWrapper } from "../uri-resolution";

import { Result } from "@polywrap/result";

// $start: UriResolver.ts

/** Options required for URI resolution. */
export interface TryResolveUriOptions {
  /** The Wrapper's URI */
  uri: Uri;

  /** A URI resolution context */
  resolutionContext?: IUriResolutionContext;
}

/** An entity capable of resolving a wrap URI, typically by using an IUriResolver implementation */
export interface UriResolverHandler<TError = undefined> {
  /**
   * Resolve a URI to a wrap package, a wrapper, or a uri
   *
   * @param options - TryResolveUriOptions
   * @returns A Promise with a Result containing either a wrap package, a wrapper, or a URI if successful
   */
  tryResolveUri(
    options?: TryResolveUriOptions
  ): Promise<Result<UriPackageOrWrapper, TError>>;
}

// $end

File: packages/rs/core/src/uri-resolver.rs
use uri::{ Uri };
use uri_resolution::{ UriResolutionContext };

struct TryResolveUriOptions {
  uri: Uri;
  resolution_context: Option<UriResolutionContext>;
}

/*
TODO
interface UriResolverHandler {
  tryResolveUri(
    options?: TryResolveUriOptions
  ) -> Future<Result<UriPackageOrWrapper, Error>>;
}
*/
